spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/gson/GsonAutoConfiguration.java

```diff
@@ -16,28 +16,176 @@
 
 package org.springframework.boot.autoconfigure.gson;
 
+import com.google.gson.FieldNamingPolicy;
 import com.google.gson.Gson;
 
+import com.google.gson.GsonBuilder;
+import com.google.gson.LongSerializationPolicy;
 import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+
+import org.springframework.context.annotation.Primary;
+import org.springframework.core.Ordered;
+import org.springframework.util.CollectionUtils;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.List;
 
 /**
  * {@link EnableAutoConfiguration Auto-configuration} for Gson.
  *
  * @author David Liu
+ * @author Ivan Golovko
  * @since 1.2.0
  */
 @Configuration
 @ConditionalOnClass(Gson.class)
 public class GsonAutoConfiguration {
 
-	@Bean
-	@ConditionalOnMissingBean
-	public Gson gson() {
-		return new Gson();
-	}
+    @Configuration
+    @ConditionalOnClass({Gson.class})
+    static class GsonConfiguration {
+
+        @Bean
+        @Primary
```

Comment:
Marking it primary has nothing to do with the configuration part this PR claims. Please restrict the PR to the configuration part.